pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,233 @@
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+
+	// Try to patch, if it fails then try to create
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		setConnectMeta(&cfg, sg.Server.instanceID, con)
+		return cfg
+	})
+	// TODO return err from Patch through Get
+	if err == nil {
+		return
+	} else if !errors.IsNotFound(err) && err.Error() != "item not found" {
+		adsLog.Warnf("updating auto-registered WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+	}
+
+	// No WorkloadEntry, create one using fields from the associated WorkloadGroup
+	groupCfg := sg.Store.Get(gvk.WorkloadGroup, proxy.Metadata.AutoRegisterGroup, proxy.Metadata.Namespace)
+	if groupCfg == nil {
+		adsLog.Warnf("auto-registration of %v failed: cannot find WorkloadGroup %s/%s", proxy.ID, proxy.Metadata.Namespace, proxy.Metadata.AutoRegisterGroup)
+		return
+	}
+	entry := workloadEntryFromGroup(entryName, proxy, groupCfg)
+	setConnectMeta(entry, sg.Server.instanceID, con)
+	_, err = sg.Store.Create(*entry)
+	if err != nil {
+		adsLog.Errorf("auto-registration of %v failed: error creating WorkloadEntry: %v", proxy.ID, err)
+	}
+}
+
+func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
+	// check if the WE already exists, update the status
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
+	}
+	// unset controller, set disconnect time
+	_, err := sg.Store.Patch(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace, func(cfg config.Config) config.Config {
+		delete(cfg.Annotations, WorkloadControllerAnnotation)
+		cfg.Annotations[DisconnectedAtAnnotation] = strconv.FormatInt(time.Now().UnixNano(), 10)
+		return cfg
+	})
+	if err != nil {
+		adsLog.Warnf("disconnect: failed patching WorkloadEntry %s/%s: %v", proxy.Metadata.Namespace, entryName, err)
+		return
+	}
+
+	ns := proxy.Metadata.Namespace
+	sg.cleanupQueue.PushDelayed(func() error {
+		wle := sg.Store.Get(gvk.WorkloadEntry, entryName, ns)
+		if wle == nil {
+			return nil
+		}
+		sg.cleanupEntry(*wle)
+		return nil
+	}, features.WorkloadEntryCleanupGracePeriod)
+}
+
+// periodicWorkloadEntryCleanup checks lists all WorkloadEntry
+func (sg *InternalGen) periodicWorkloadEntryCleanup(stopCh <-chan struct{}) {
+	ticker := time.NewTicker(10 * features.WorkloadEntryCleanupGracePeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			wles, err := sg.Store.List(gvk.WorkloadEntry, metav1.NamespaceAll)
+			if err != nil {
+				continue
+			}
+			for _, wle := range wles {
+				wle := wle
+				sg.cleanupQueue.Push(func() error {
+					sg.cleanupEntry(wle)
+					return nil
+				})
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (sg *InternalGen) cleanupEntry(wle config.Config) {
+	if !shouldCleanupEntry(wle) {
```

Comment:
should we filter before adding to the queue to keep the queue smaller?

Reply:
For the periodic cleanup, that'll work. Still need to call this prior to `cleanupEntry` in the delayed queue tasks.